src/rgw/rgw_op.cc

```diff
@@ -5234,6 +5234,56 @@ int RGWCopyObj::verify_permission(optional_yield y)
 					      RGW_PERM_READ)) { 
 	  return -EACCES;
 	}
+
+  // Check for tagging permissions
+  bool is_put_tagging_permission = false, is_get_tagging_permission = false, without_tagging_directive = false;
+  bool is_tagging_directive_copy = false , is_tagging_directive_replace = false;
+  auto tagging_drctv = s->info.env->get("HTTP_X_AMZ_TAGGING_DIRECTIVE");
+  if (tagging_drctv) {
+    if (strcasecmp(tagging_drctv, "COPY") == 0)
+      is_tagging_directive_copy = true;
+    else if (strcasecmp(tagging_drctv, "REPLACE") == 0)
+      is_tagging_directive_replace = true;
+  } else {
+      without_tagging_directive = true;
+    }
+
+  auto put_tagging_permission = eval_identity_or_session_policies(this,s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3PutObjectTagging,
+                                                                  obj_arn);
+  if (put_tagging_permission == Effect::Allow )
+    is_put_tagging_permission = true;
+
+  auto get_tagging_permission = eval_identity_or_session_policies(this,s->iam_user_policies,
+                                                                  s->env,
+                                                                  rgw::IAM::s3GetObjectTagging,
+                                                                  obj_arn);
+  if (get_tagging_permission == Effect::Allow )
+    is_get_tagging_permission = true;
+
+  op_ret = s->src_object->get_obj_attrs(s->yield, this);
+  if (op_ret == 0) {
+      attrs = s->src_object->get_attrs();
+      auto tags = attrs.find(RGW_ATTR_TAGS);
+      if (tags != attrs.end()) {
+        //tags present on source object
+        if (is_put_tagging_permission && !is_get_tagging_permission) {
+          if(is_tagging_directive_copy || without_tagging_directive)
+            return -EACCES;
+        } else if (!is_put_tagging_permission && is_get_tagging_permission) {
+          return -EACCES;
+        } else if (!is_put_tagging_permission && !is_get_tagging_permission) {
+          return -EACCES;
+        }
+      } else {
+        //tags not present on source obj
+        if (!is_put_tagging_permission && is_get_tagging_permission) {
+            if(is_tagging_directive_replace)
+            return -EACCES;
+        }
+      }
```

Comment:
where is case 4 , where getObjectTagging & putObjectTagging permissions are granted and code here needs to check that !?
> 4] Apply tags on source object and give putObject, get-object-tagging, get-object,put-objectTagging permissions in IAM policy

For that case, check can be
> if ( tagging_permission == GET_TAGGING_PERM | PUT_TAGGING_PERM  )

Reply:
No need to handle case 4 explicitly as all the required tagging permissions are there so it will not block but allow copy operation.

Reply:
Well tagging_permission checks above is explicitly against PUT_TAGGING_PERM(value=1) & GET_TAGGING_PERM (value=2) separately, and not taking care of the case when both permissions(value=3) available.